beacon-chain/sync/subscriber_beacon_blocks.go

```diff
@@ -2,46 +2,40 @@ package sync
 
 import (
 	"context"
-	"encoding/base64"
-	"time"
 
 	"github.com/gogo/protobuf/proto"
-	"github.com/karlseguin/ccache"
 	"github.com/prysmaticlabs/go-ssz"
 	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
 	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
 	"github.com/prysmaticlabs/prysm/shared/bytesutil"
 )
 
-// pendingBlocks cache.
-// TODO(3147): This cache value should represent a doubly linked list or some management to
-// process pending blocks after a link to the canonical chain is found.
-var pendingBlocks = ccache.New(ccache.Configure())
-
 func (r *RegularSync) beaconBlockSubscriber(ctx context.Context, msg proto.Message) error {
 	block := msg.(*ethpb.BeaconBlock)
 
 	headState := r.chain.HeadState()
-
 	// Ignore block older than last finalized checkpoint.
 	if block.Slot < helpers.StartSlot(headState.FinalizedCheckpoint.Epoch) {
 		log.Debugf("Received a block that's older than finalized checkpoint, %d < %d",
 			block.Slot, helpers.StartSlot(headState.FinalizedCheckpoint.Epoch))
 		return nil
 	}
 
-	// Handle block when the parent is unknown.
+	blockRoot, err := ssz.SigningRoot(block)
+	if err != nil {
+		log.Errorf("Could not sign root block: %v", err)
+		return nil
+	}
+	// Ignore block already in pending blocks cache
+	if r.seenPendingBlocks[blockRoot] {
```

Comment:
This should go in the validation. We wouldn't want to propagate the same pending block multiple times.

Reply:
Done